staging/src/k8s.io/apiserver/pkg/server/read_write_deadline_test.go

```diff
@@ -0,0 +1,522 @@
+		//   d) the handler flushes the ResponseWriter object before request times out: No
+		// observation:
+		//  the timeout filter detects that the context of the request has exceeded its
+		//  deadline, and the the following takes place:
+		//   - it marks the ResponseWriter object as timeout=true, so any further attempt
+		//   to write to it will yield an 'http: Handler timeout' error
+		//   - it can't send '504 GatewayTimeout' to the client since the ResponseWriter
+		//   object has already been written to, so it panics with 'net/http: abort Handler' error
+		//  at the same time, the net/http client also detects that the context of the
+		//  client-side request has exceeded its deadline, and so it aborts with a
+		//  'context deadline exceeded' error.
+		// NOTE: although the client is most likely to receive the context deadline error
+		// first due to the roundtrip time added to the arrival of the error from
+		// the server, nevertheless it could cause flakes in CI due to overload, so we
+		// need to check for either error to be flake free.
+		{
+			setup: setup{
+				name:                        "client specifies a timeout",
+				clientTimeout:               time.Second,                 // b
+				handlerWritesBeforeTimeout:  true,                        // c
+				handlerFlushesBeforeTimeout: false,                       // d
+				serverReqTimeout:            wait.ForeverTestTimeout * 2, // this should not be in effect
+
+				// twice the request timeout so it can withstand flakes in CI
+				waiter: &waitWithDuration{after: 2 * time.Second},
+			},
+			expectations: map[string]expectation{
+				// expectation:
+				//  - client: receives either a context.DeadlineExceeded error from its transport
+				//    or it receives the error from the server
+				//  - server: Write to the ResponseWriter will yields an 'http: Handler timeout'
+				"HTTP/2.0": {
+					clientErr: wantEitherOr{
+						err:      context.DeadlineExceeded,
+						contains: "stream error: stream ID 1; INTERNAL_ERROR; received from peer",
+					},
+					clientStatusCodeExpected: 0,
+					clientRespBodyReadErr:    wantNoError{},
+					handlerWriteErr:          wantError{http.ErrHandlerTimeout},
+				},
+				// expectation:
+				//  - client: receives either a context.DeadlineExceeded error from its
+				//    transport, or it receives the error from the server
+				//  - server: Write to the ResponseWriter will yields an 'http: Handler timeout'
+				"HTTP/1.1": {
+					clientErr: wantEitherOr{
+						err:      context.DeadlineExceeded,
+						contains: "EOF",
+					},
+					clientStatusCodeExpected: 0,
+					clientRespBodyReadErr:    wantNoError{},
+					handlerWriteErr:          wantError{http.ErrHandlerTimeout},
+				},
+			},
+		},
+	}
+	for _, test := range tests {
+		for _, proto := range []string{"HTTP/1.1", "HTTP/2.0"} { // every test is run for both http/1x and http/2.0
+			t.Run(fmt.Sprintf("%s/%s", test.setup.name, proto), func(t *testing.T) {
```

Comment:
these sub-tests can't be run in parallel using `t.Parallel'  when we run these tests with the feature gate `PerHandlerReadWriteTimeout` enabled/disabled

Reply:
I will implement the pattern for parallel subtests - ref: https://github.com/kubernetes/kubernetes/pull/122924